File: frontend/types/habit.ts
export interface HabitResponse {
  id: number;
  name: string;
  category: string;
  reminder: string;
  streak: number;
  longestStreak: number;
  completion: number;
  missedDays: number;
  doneToday: boolean;
  paused: boolean;
  progress: number;
  week: boolean[];
}

export type CreateHabitPayload = Omit<
  HabitResponse,
  "id" | "streak" | "longestStreak" | "completion" | "missedDays" | "doneToday" | "progress" | "week"
>;

export type UpdateHabitPayload = Partial<HabitResponse>;

File: app/schemas/habit.py
from typing import Optional

from pydantic import BaseModel


class HabitCreate(BaseModel):
    name: str
    category: str = "General"
    reminder: str | None = None
    paused: bool = False
    frequency: str = "Daily"
    target_count: int = 1


class HabitUpdate(BaseModel):
    name: Optional[str] = None
    category: Optional[str] = None
    reminder: Optional[str] = None
    paused: Optional[bool] = None
    frequency: Optional[str] = None
    target_count: Optional[int] = None
    # Convenience flag the frontend uses to toggle today's completion;
    # mapped to real HabitLog rows + streak bookkeeping server-side.
    doneToday: Optional[bool] = None


class HabitResponse(BaseModel):
    id: int
    name: str
    category: str
    reminder: str | None = None
    streak: int
    longestStreak: int
    completion: int
    missedDays: int
    doneToday: bool
    paused: bool
    progress: int
    week: list[bool]

    class Config:
        from_attributes = True
